### PR TITLE
Add MambaForSequenceClassification

### PR DESCRIPTION
**Pull request by mjschock**
# What does this PR do?

Adds `MambaForSequenceClassification` for sequence classification with the `MambaModel` backbone.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker thanks for your work bringing in Mamba! I'm wondering if there's any objection to adding a MambaForSequenceClassification model?  I followed the template example as best as I could and happy to continue with adding the new test and finishing it up.

### CODE REVIEW

File: src/transformers/models/mamba/modeling_mamba.py [hunk 1]
```diff
@@ -679,3 +705,109 @@ def forward(
...
+@add_start_docstrings(
+    """Mamba Model backbone with a sequence classification/regression head on top (a linear layer on top of
+    the pooled output) e.g. for GLUE tasks.""",
+    MAMBA_START_DOCSTRING,
+)
+class MambaForSequenceClassification(MambaPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+        self.num_labels = config.num_labels
+        self.backbone = MambaModel(config)
```
Comment (amyeroberts) [inline src/transformers/models/mamba/modeling_mamba.py:740]: I wouldn't call this a `backbone` - this has the implication it's extracting features, which doesn't really match the intention here

```suggestion
        self.model = MambaModel(config)
```

**Review by amyeroberts — COMMENTED** [review]
Hi @mjschook, thanks for opening this PR and adding this! 

AFAICT, these changes seem reasonable to me. @ArthurZucker is off for a week - so let's wait for him to come back to confirm if there's any reason for not adding this to Mamba. 

A few things that will need to be added: 
* Tests for the model i.e. equivalent to `create_and_check_mamba_model` and the model should be added to `all_model_classes` 
* The model needs to be documented in `mamba.md` 
* All the tests in the CI should be passing

File: src/transformers/models/mamba/modeling_mamba.py (hunk 1 above)
Comment (mjschock) [reply src/transformers/models/mamba/modeling_mamba.py:740]: I think it needs to be `backbone` in this case b/c that's the `base_model_prefix` and the PyTorch weights use the same convention.

File: src/transformers/models/mamba/modeling_mamba.py [hunk 2]
```diff
@@ -679,3 +709,113 @@ def forward(
...
+        self.classifier = MambaClassificationHead(config)
+
+        for param in self.base_model.parameters():
+            param.requires_grad = False
```
Comment (mjschock) [inline src/transformers/models/mamba/modeling_mamba.py:749]: I'm not sure whether we actually want to freeze the params for the base model here, but I found there's [a test specific to sequence classification that expects all the unfrozen params to be initialized in the range `[0.0, 1.0]` ](https://app.circleci.com/pipelines/github/huggingface/transformers/87066/workflows/ee804f42-0535-452c-8ee6-38df520712e3/jobs/1130735/parallel-runs/0/steps/0-115) and the initialized values for the mixer don't satisfy that assertion.

So... I froze them and made sure [the classification head params were initialized to satisfiy the test](https://github.com/huggingface/transformers/pull/29552/files#diff-e1d4758c08973fdac2c23a8a3710872d943ce8509035205da4a681bc4dcaf1c3R365-R368).  It makes intuitive sense to me to freeze them in the case of transfer learning for this task and I did confirm that running LoRA PEFT with `target_modules=["x_proj", "embeddings", "in_proj", "out_proj"]` and `task_type=TaskType.SEQ_CLS` does unfreeze the target modules so it appears to work fine, but not sure if we want to force them to be frozen by default. 

Anyway, happy to adjust if there's a better practice to follow here.

**Review by ArthurZucker — COMMENTED** [review]
Hey! We usually do the following before merging such PRs:
1. Create a feature request issue to add this new class
2. Wait until the community picks this up
3. Wait until there are actually pretrained checkpoints released by the community or the authors. 

As is, this does not really help anyone as it can be easily implemented by anyone that wants to train a model no?

**Comment by mjschock** [comment]
Thanks for the feedback @ArthurZucker - I'll close it since I went another direction, using prompt tuning instead.  I'll keep the process you laid out in mind for the future. =)

**Pull request closed without merging**